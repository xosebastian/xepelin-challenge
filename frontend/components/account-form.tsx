"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";

import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { useForm } from "react-hook-form";
import { useDispatch } from "react-redux";
import { AppDispatch, useAppDispatch, useAppSelector } from "@/redux/store";
import { create } from "@/redux/features/account-slice";

const formSchema = z.object({
  accountName: z.string().min(2, {
    message: "The name should have at least 2 characters.",
  }),
  accountNumber: z
    .string()
    .min(5, {
      message: "The account number should have at least 5 digits.",
    })
    .regex(/^\d+$/, {
      message: "The account number should only contain numbers.",
    }),
  balance: z
    .string()
    .max(4, {
      message: "The balance should have at most 4 digits.",
    })
    .regex(/^\d+(\.\d{1,2})?$/, {
      message: "The balance should be a number with up to 2 decimal places.",
    })
    .transform((value) => parseFloat(value)),
});

export function AccountForm() {
  const dispatch = useAppDispatch();
  const account = useAppSelector((state) => state.account);

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      accountName: "",
      accountNumber: "",
      balance: 0,
    },
  });

  function onSubmit(values: z.infer<typeof formSchema>) {
    // Do something with the form values.
    // âœ… This will be type-safe and validated.

    const { accountName, accountNumber, balance } = values;

    dispatch(
      create({
        id: "1",
        accountName,
        accountNumber,
        balance,
      })
    );
  }

  if (account.id)
    return (
      <div className="flex flex-col space-y-1.5 p-6">
        <h3 className="text-2xl font-semibold leading-none tracking-tight">
          My Account
        </h3>
        <ul className="flex flex-col space-y-1.5 pt-5">
          <li className="text-1xl font-semibold leading-none tracking-tight">
            id: {account.id}
          </li>
          <li className="text-1xl font-semibold leading-none tracking-tight">
            name: {account.accountName}
          </li>
          <li className="text-1xl font-semibold leading-none tracking-tight">
            number: {account.accountNumber}
          </li>
          <li className="text-1xl font-semibold leading-none tracking-tight">
            balance: {account.balance}
          </li>
        </ul>
      </div>
    );

  return (
    <>
      <div className="flex flex-col space-y-1.5 p-6">
        <h3 className="text-2xl font-semibold leading-none tracking-tight">
          Create Account
        </h3>
      </div>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-2 p-5">
          <FormField
            control={form.control}
            name="accountName"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Name</FormLabel>
                <FormControl>
                  <Input placeholder="Enter your account name" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="accountNumber"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Number</FormLabel>
                <FormControl>
                  <Input
                    maxLength={5}
                    placeholder="Enter your account number"
                    {...field}
                  />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="balance"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Balance</FormLabel>
                <FormControl>
                  <Input type="number" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <Button type="submit">Create</Button>
        </form>
      </Form>
    </>
  );
}
