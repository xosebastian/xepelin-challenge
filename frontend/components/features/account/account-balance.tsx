"use client";
import { useAppDispatch, useAppSelector } from "@/redux/store";
import axios, { AxiosResponse } from "axios";
import { useEffect } from "react";
import { update } from "@/redux/features/account-slice";

export function AccountBalance() {
  const accountId = useAppSelector((state) => state.account.id);
  const balance = useAppSelector((state) => state.account.balance);
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (!accountId) return;

    const getBalance = async () => {
      try {
        const {
          data: { balance },
        }: AxiosResponse<{ balance: string }> = await axios.get(
          `/api/account/balance/${accountId}`,
          {}
        );

        console.log(balance);

        dispatch(
          update({
            balance: parseFloat(balance),
          })
        );
      } catch (error) {
        console.error(error);
      }
    };

    getBalance();
  }, [accountId, dispatch]);

  return (
    <div className="flex flex-col space-y-1.5 p-6">
      <h3 className="text-2xl font-semibold leading-none tracking-tight" data-testid="balance">
        $ {balance}
      </h3>
    </div>
  );
}
