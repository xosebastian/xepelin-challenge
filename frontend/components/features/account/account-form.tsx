"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";

import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { useForm } from "react-hook-form";

export const formAccountSchema = z.object({
  accountName: z.string().min(2, {
    message: "The name should have at least 2 characters.",
  }),
  accountNumber: z
    .string()
    .min(10, {
      message: "The account number should have at least 10 digits.",
    })
    .regex(/^\d+$/, {
      message: "The account number should only contain numbers.",
    }),
  balance: z
    .string()
    .max(5, {
      message: "The balance should have at most 5 digits.",
    })
    .regex(/^\d+(\.\d{1,2})?$/, {
      message: "The balance should be a number with up to 2 decimal places.",
    })
    .transform((value) => parseFloat(value)),
});

interface Props {
  onSubmit: (values: z.infer<typeof formAccountSchema>) => void;
  onChange?: () => void;
}

export function AccountForm({ onSubmit, onChange }: Props) {
  const form = useForm<z.infer<typeof formAccountSchema>>({
    resolver: zodResolver(formAccountSchema),
    defaultValues: {
      accountName: "",
      accountNumber: "",
      balance: 0,
    },
  });

  return (
    <Form {...form}>
      <form
        onSubmit={form.handleSubmit(onSubmit)}
        className="space-y-2 p-5"
        onChange={onChange}
      >
        <FormField
          control={form.control}
          name="accountName"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Name</FormLabel>
              <FormControl>
                <Input placeholder="Enter your account name" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="accountNumber"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Number</FormLabel>
              <FormControl>
                <Input
                  maxLength={10}
                  placeholder="Enter your account number"
                  {...field}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="balance"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Balance</FormLabel>
              <FormControl>
                <Input type="number" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <Button role="button" type="submit">
          Create
        </Button>
      </form>
    </Form>
  );
}
