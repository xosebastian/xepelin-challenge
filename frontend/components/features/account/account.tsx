"use client";

import * as z from "zod";

import { Button } from "@/components/ui/button";

import { useAppDispatch, useAppSelector } from "@/redux/store";
import { create, clear } from "@/redux/features/account-slice";
import axios, { AxiosResponse, isAxiosError } from "axios";
import { useState } from "react";
import { Alert, AlertDescription, AlertTitle } from "../../ui/alert";
import { useToast } from "../../ui/use-toast";
import { AccountForm, formAccountSchema } from "./account-form";

export function Account() {
  const dispatch = useAppDispatch();
  const accountId = useAppSelector((state) => state.account.id);
  const [error, setError] = useState<string | null>(null);
  const { toast } = useToast();

  async function onSubmit(values: z.infer<typeof formAccountSchema>) {
    const { accountName, accountNumber, balance } = values;

    try {
      const {
        data: { accountId },
      }: AxiosResponse<{ accountId: string }> = await axios.post(
        "/api/account/create",
        {
          accountName,
          accountNumber,
          balance,
        }
      );

      dispatch(
        create({
          id: accountId,
        })
      );

      toast({
        title: "Account created",
        description: `Account ${accountName} created with id ${accountId}`,
      });
    } catch (error: any) {
      const message = isAxiosError(error)
        ? error?.response?.data.message
        : error?.message ?? "Something went wrong";

      setError(message);
    }
  }

  return (
    <>
      {accountId ? (
        <div className="flex flex-col space-y-1.5 p-6">
          <h3 className="text-2xl font-semibold leading-none tracking-tight">
            My Account
          </h3>
          <p className="text-sm text-muted-foreground">id: {accountId}</p>
          <Button
            onClick={() => {
              dispatch(clear());
            }}
          >
            Create Another
          </Button>
        </div>
      ) : (
        <>
          <div className="flex flex-col space-y-1.5 p-6">
            <h3 className="text-2xl font-semibold leading-none tracking-tight">
              Create Account
            </h3>
          </div>
          <AccountForm onSubmit={onSubmit} onChange={() => setError(null)} />
          {error && (
            <Alert variant={"destructive"} className="mt-5">
              <AlertTitle>Ups!</AlertTitle>
              <AlertDescription>{error}</AlertDescription>
            </Alert>
          )}
        </>
      )}
    </>
  );
}
