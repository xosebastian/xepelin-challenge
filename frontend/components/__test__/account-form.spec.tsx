import { render, screen, fireEvent, act } from '@testing-library/react';
import { AccountForm } from '../features/account/account-form';


describe('AccountForm', () => {
  let onSubmitMock = jest.fn();

  beforeEach(() => {
    render(<AccountForm onSubmit={onSubmitMock} />);
  });

  it("should render the Xeplin Bank link", () => {
    const { container } = render(<AccountForm onSubmit={onSubmitMock} />);
    expect(container).toMatchSnapshot();
  });

  it('should render account name input', () => {
    expect(screen.getByTestId('accountName')).toBeDefined();
  });

  it('should render account number input', () => {
    expect(screen.getByTestId('accountNumber')).toBeDefined();
  });

  it('should render balance input', () => {
    expect(screen.getByTestId('balance')).toBeDefined();
  });

  it('should render create button', () => {
    expect(screen.getByRole('button', { name: 'Create' })).toBeDefined();
  });

  it('should call onSubmit when the form is submitted', async () => {
    await act(async () => {
      fireEvent.change(screen.getByTestId('accountName'), { target: { value: 'Test Name' } });
      fireEvent.change(screen.getByTestId('accountNumber'), { target: { value: '1234567890' } });
      fireEvent.change(screen.getByTestId('balance'), { target: { value: 100 } });
      fireEvent.click(screen.getByRole('button', { name: 'Create' }));
    });

    expect(onSubmitMock).toHaveBeenCalledTimes(1);
  });
});
