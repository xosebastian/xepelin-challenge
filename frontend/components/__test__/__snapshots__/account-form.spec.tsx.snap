// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm should render the Xeplin Bank link 1`] = `
<div>
  <form
    class="space-y-2 p-5"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r3:-form-item"
      >
        Name
      </label>
      <input
        aria-describedby=":r3:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="accountName"
        id=":r3:-form-item"
        name="accountName"
        placeholder="Enter your account name"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r4:-form-item"
      >
        Number
      </label>
      <input
        aria-describedby=":r4:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="accountNumber"
        id=":r4:-form-item"
        maxlength="10"
        name="accountNumber"
        placeholder="Enter your account number"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r5:-form-item"
      >
        Balance
      </label>
      <input
        aria-describedby=":r5:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="balance"
        id=":r5:-form-item"
        name="balance"
        type="number"
        value="0"
      />
    </div>
    <button
      class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      data-testid="create"
      role="button"
      type="submit"
    >
      Create
    </button>
  </form>
</div>
`;
